File: mq/rabbitmq/py/topic/subscriber_all.py
#!/usr/bin/env python3
#
# Copyright (C) 2019 jayson
#
import json
import contextlib
import utils.logging
from config import RABBITMQ_MASTER
from utils.connection import get_exchange_connection
from pika.adapters.blocking_connection import BlockingChannel
from pika.spec import Basic, BasicProperties


@contextlib.contextmanager
def consumer_factory(log_file: str):
    with open(log_file, 'a') as w:
        def __consumer(ch: BlockingChannel, method: Basic.Deliver, properties: BasicProperties, body: bytes):
            data = json.loads(body)
            message = '{:<16}@{:<16}@{:<4}@{:<20}: {}'.format(
                data['from'], data['producer'], data['index'], method.routing_key, data['body'])
            w.write(message + '\n')
            utils.logging.info(message)
            ch.basic_ack(delivery_tag=method.delivery_tag)
        yield __consumer


if __name__ == '__main__':
    with get_exchange_connection(RABBITMQ_MASTER, 'logs_topic', 'topic') as channel:
        channel.queue_declare(queue='logs_topic_all', durable=False, auto_delete=True)
        channel.queue_bind(queue='logs_topic_all', exchange='logs_topic', routing_key='#')

        with consumer_factory('topic.log') as consumer:
            channel.basic_consume(queue='logs_topic_all', on_message_callback=consumer, auto_ack=False)
            channel.start_consuming()

File: mq/rabbitmq/php/src/topic/subscriber_all.php
<?php
/**
 * rabbitmq hello-world
 */
require_once __DIR__ . '/../bootstrap.php';


$connection = new \PhpAmqpLib\Connection\AMQPStreamConnection(RABBITMQ_MASTER, 5672, 'guest', 'guest');
$channel = $connection->channel();

$channel->exchange_declare('logs_topic', 'topic', false, true, false);
$channel->queue_declare('logs_topic_all', false, false, false, true);
$channel->queue_bind('logs_topic_all', 'logs_topic', '#');

$channel->basic_qos(null, 1, true);
$channel->basic_consume('logs_topic_all', null, false, false, false, false,
    function(\PhpAmqpLib\Message\AMQPMessage $message) use ($channel) {
        $data = json_decode($message->body, true);
        echo sprintf("%-12s %-12s %-8s %-20s: %s\n", $data['from'], $data['producer'], $data['index'],
            $message->delivery_info['routing_key'], $data['body']);

        if (random_int(0, 20) === 1) {
            $channel->basic_reject($message->getDeliveryTag(), true);
            return;
        }
        $channel->basic_ack($message->getDeliveryTag());
    }
);

while ($channel->is_consuming()) {
    $channel->wait();
}
